### PR TITLE
Extend PotGenerator to improve plural forms output

### LINKED ISSUE #121: POT file generated with this command produces an error with Transifex service related to plural forms
This is a fairly specific issue, but hopefully someone knows something (if the problem is on our end, feel free to close).

We've been using POEdit to (re)generate the POT file for [Pressbooks](https://github.com/pressbooks/pressbooks) and we use [Transifex](https://transifex.com) to manage translations. I have been trying out the wp-cli `make-pot` command and I get this error from the Transifex CLI when trying to upload the resulting POT file:

```
Error received from server: Your source file is not a POT file and the translation file you're using has more than two plurals which is not supported.
```

I ran the command with these options from the plugin root:

```
wp i18n make-pot . languages/_pressbooks.pot --exclude="vendor,symbionts" --headers='{"Report-Msgid-Bugs-To":"https://github.com/pressbooks/pressbooks/"}'
```

I note that in the wp-cli generated file, the second `msgstr` is missing from plural strings:

```diff
#: inc/class-cloner.php:2045 inc/modules/import/wordpress/class-wxr.php:337
#, php-format
msgid "%s part"
msgid_plural "%s parts"
msgstr[0] ""
- msgstr[1] ""
```

Is this expected?

Thanks in advance for any information that might lead to resolution!

### PR DESCRIPTION
**Pull request by swissspidy**
Fixes #121.

Note: Might need some more tests because this adds quite a lot of code from the upstream library.

### CODE REVIEW

**Comment by greatislander** [comment]
@swissspidy I didn't see this— I'll test it with our project and see if it resolves the issue. Thanks!

File: src/PotGenerator.php [hunk 1]
```diff
@@ -35,12 +35,105 @@ public static function setCommentBeforeHeaders( $comment ) {
...
+			}
+
+			if ( $translation->hasExtractedComments() ) {
+				foreach ( $translation->getExtractedComments() as $comment ) {
+					$lines[] = '#. ' . $comment;
+				}
+			}
+
+			foreach ( $translation->getReferences() as $reference ) {
+				$lines[] = '#: ' . $reference[0] . ( $reference[1] !== null ? ':' . $reference[1] : null );
```
Comment (schlessera) [inline src/PotGenerator.php]: I'd say `''` is a better no-op here when concatenating:
```suggestion
				$lines[] = '#: ' . $reference[0] . ( $reference[1] !== null ? ':' . $reference[1] : '' );
```

File: src/PotGenerator.php [hunk 2]
```diff
@@ -35,12 +35,105 @@ public static function setCommentBeforeHeaders( $comment ) {
...
+
+		foreach ( $translations as $translation ) {
+			/** @var \Gettext\Translation $translation */
+			if ( $translation->hasComments() ) {
+				foreach ( $translation->getComments() as $comment ) {
+					$lines[] = '# ' . $comment;
+				}
+			}
+
+			if ( $translation->hasExtractedComments() ) {
```
Comment (schlessera) [inline src/PotGenerator.php]: This is not really needed, as you'll get an empty array back if not. Iterating over an empty array just does nothing anyway.

File: src/PotGenerator.php [hunk 3]
```diff
@@ -35,12 +35,105 @@ public static function setCommentBeforeHeaders( $comment ) {
...
+				}
+			} else {
+				self::addLines( $lines, $prefix . 'msgstr', $translation->getTranslation() );
+			}
+
+			$lines[] = '';
 		}
 
-		return $result . parent::toString( $translations, $options );
+		return implode("\n", $lines);
```
Comment (schlessera) [inline src/PotGenerator.php]: CS:
```suggestion
		return implode( "\n", $lines );
```

File: src/PotGenerator.php [hunk 4]
```diff
@@ -35,12 +35,105 @@ public static function setCommentBeforeHeaders( $comment ) {
...
+
+		foreach ( $translations->getHeaders() as $name => $value ) {
+			$lines[] = sprintf( '"%s: %s\\n"', $name, $value );
+		}
+
+		$lines[] = '';
+
+		foreach ( $translations as $translation ) {
+			/** @var \Gettext\Translation $translation */
+			if ( $translation->hasComments() ) {
```
Comment (schlessera) [inline src/PotGenerator.php]: This is not really needed, as you'll get an empty array back if not. Iterating over an empty array just does nothing anyway.

File: src/PotGenerator.php [hunk 5]
```diff
@@ -35,12 +35,105 @@ public static function setCommentBeforeHeaders( $comment ) {
...
 	public static function toString( Translations $translations, array $options = [] ) {
-		$result = '';
+		$lines   = static::$comments_before_headers;
+		$lines[] = 'msgid ""';
+		$lines[] = 'msgstr ""';
 
-		if ( ! empty( static::$comments_before_headers ) ) {
-			$result = implode( "\n", static::$comments_before_headers ) . "\n";
+		$plural_form = $translations->getPluralForms();
+		$plural_size = is_array( $plural_form ) ? ( $plural_form[0] - 1 ) : 1;
```
Comment (schlessera) [inline src/PotGenerator.php]: This needs a comment to explain what happens and especially what `$plural_form[0] - 1` means.

**Comment by schlessera** [comment]
The code is not really pretty because there's so much static access going on in the original library, but I guess that can't be easily helped.

**Comment by connerbw** [comment]
Can this be merged soon? I use it.

**Comment by connerbw** [comment]
If this isn't going to be merged soon, can someone sync `121-plural-forms` with `master`? It's ~23 commits behind, there are fixes this branch could benefit from. (For example; #125)  Then I can simply keep using `composer require wp-cli/i18n-command:dev-121-plural-forms`

**Comment by schlessera** [comment]
@swissspidy I'm fine with merging this now. Do you feel good about this as well?

Regarding the tests, I would assume we'll notice when things go wrong because of the Behat tests anyway, even without precise unit tests?

**Comment by swissspidy** [comment]
Sure. We can do some housekeeping here at some point in the future I guess.

**Comment by schlessera** [comment]
@connerbw This is merged into `master` now.

**Pull request merged**